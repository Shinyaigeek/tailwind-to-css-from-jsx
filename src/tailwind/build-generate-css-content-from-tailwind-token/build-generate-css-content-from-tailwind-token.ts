import { tailwindToCSS, twi as plainTwi } from "npm:tw-to-css";
import type { Config } from "npm:tailwindcss";
import { createErr, createOk, Result } from "npm:option-t/plain_result";

export const buildGenerateCSSContentFromTailwindToken: (
  tailwindConfig?: Config,
) => (twToken: string) => Result<string, Error> = function (tailwindConfig) {
  const twi = (() => {
    if (tailwindConfig) {
      return tailwindToCSS({ config: tailwindConfig }).twi;
    } else {
      return plainTwi;
    }
  })();

  return function (twToken) {
    try {
      const generatedCSS = twi(twToken, {
        ignoreMediaQueries: false,
        minify: false,
      });
      if (generatedCSS === "") {
        return createErr(
          new Error(`Failed to generate CSS from Tailwind token ${twToken}`),
        );
      }
      return createOk(generatedCSS);
    } catch (e) {
      return createErr(
        new Error(
          `Failed to generate CSS from Tailwind token ${twToken}: ${e}`,
        ),
      );
    }
  };
};
