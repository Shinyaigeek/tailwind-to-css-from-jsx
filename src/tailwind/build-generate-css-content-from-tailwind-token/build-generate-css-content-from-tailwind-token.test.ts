import { buildGenerateCSSContentFromTailwindToken } from "./build-generate-css-content-from-tailwind-token.ts";
import { assertEquals } from "jsr:@std/assert";
import { isOk } from "option-t/plain_result";
import { createAssertSnapshot } from "jsr:@std/testing/snapshot";
import { stripAnsiCode } from "@std/fmt/colors";
import { unwrapOk } from "option-t/plain_result";

const assertSnapshot = createAssertSnapshot({
  dir: ".snaps",
});

const assertMonochromeSnapshot = createAssertSnapshot<string>(
  { serializer: stripAnsiCode },
  assertSnapshot,
);

Deno.test({
  name:
    "generated function should return CSS content from valid Tailwind token",
  fn: async (t) => {
    const generateCSSContentFromTailwindToken =
      buildGenerateCSSContentFromTailwindToken();

    const validTailwindTokens = [
      "bg-blue-500",
      "text-red-500",
      "p-4",
      "m-4",
      "rounded-lg",
      "shadow-lg",
      "hover:bg-blue-700",
      "focus:outline-none",
      "focus:ring-2",
      //  TODO: We need to re-implement CSS Content generator from scratch to support latest TailwindCSS features
      //   "focus:ring-blue-500",
      //   "focus:ring-opacity-50",
      "sm:text-center",
      "w-1/2",
      "w-[12px]",
    ];

    for (const validTailwindToken of validTailwindTokens) {
      const result = generateCSSContentFromTailwindToken(validTailwindToken);
      assertEquals(isOk(result), true);

      await assertMonochromeSnapshot(t, unwrapOk(result));
    }
  },
});
