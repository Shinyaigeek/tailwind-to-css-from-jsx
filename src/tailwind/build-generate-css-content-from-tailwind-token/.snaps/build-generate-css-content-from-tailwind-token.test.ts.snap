export const snapshot = {};

snapshot[`generated function should return CSS content from valid Tailwind token 1`] = `
background-color: rgb(59,130,246); 

`;

snapshot[`generated function should return CSS content from valid Tailwind token 2`] = `
color: rgb(239,68,68); 

`;

snapshot[`generated function should return CSS content from valid Tailwind token 3`] = `
padding: 1rem; 

`;

snapshot[`generated function should return CSS content from valid Tailwind token 4`] = `
margin: 1rem; 

`;

snapshot[`generated function should return CSS content from valid Tailwind token 5`] = `
border-radius: 0.5rem; 

`;

snapshot[`generated function should return CSS content from valid Tailwind token 6`] = `
box-shadow: 0 0 #0000, 0 0 #0000, 0 10px 15px -3px rgb(0,0,0,0.1), 0 4px 6px -4px rgb(0,0,0,0.1); 

`;

snapshot[`generated function should return CSS content from valid Tailwind token 7`] = `
background-color: rgb(29,78,216); 

`;

snapshot[`generated function should return CSS content from valid Tailwind token 8`] = `
outline: 2px solid transparent; 
outline-offset: 2px; 

`;

snapshot[`generated function should return CSS content from valid Tailwind token 9`] = `
box-shadow: 0 0 0 0px #fff,   0 0 0 calc(2px + 0px) rgb(59,130,246,0.5), 0 0 #0000; 

`;

snapshot[`generated function should return CSS content from valid Tailwind token 10`] = `
text-align: center; 

@media(min-width: 640px) {
	text-align: center; 
}

`;

snapshot[`generated function should return CSS content from valid Tailwind token 11`] = `
width: 50%; 

`;

snapshot[`generated function should return CSS content from valid Tailwind token 12`] = `
width: 12px; 

`;
